### PR TITLE
Add support for schema synchronization 

### PR DESCRIPTION
**Pull request by khanaffan**
Lock-free schema sync allows the schema to be synchronized without import/export between two or more briefcases for a given iModel. It will use `CodeService` to hold `sync-db`. 

### Todo
- [x] Allow additive changes to be synchronized between briefcases via a **sync-db**.
- [x] Allow `ImodelConsole` to pull/push syncdb.
- [x] Add `PRAGMA checksum(ec_schema|ec_map|db_schema)` to compute sha1 checksum of respective facet.
- [x] Able to create syncdb from existing iModel. 
- [x] Add data version to avoid unnecessary updates when pulling changes from sync db.
- [x] Support conflicting changes by two or more users.
- [X] Add/Update ECDb Unit test to verify all existing schema upgrade test work with a shared channel #244
- [X] Add api and test on typescript side. (dgndb.test.ts)
- [X] Add support for cloud sqlite on itwin.js side and necessary test.  [#5620](https://github.com/iTwin/itwinjs-core/pull/5620)

itwinjs-core: https://github.com/iTwin/itwinjs-core/tree/afk/shared-schema-channel-support

### CODE REVIEW

**Comment by pmconne** [comment]
What are your intentions for the remaining TODOs? You mention itwinjs-core changes but you haven't changed the API.

**Comment by khanaffan** [comment]
> * or
I changed it back to draft.  I thought it was a different PR. This one still needs more work including #186 merged into this branch.

File: iModelCore/BeSQLite/PublicAPI/BeSQLite/BeSQLite.h [hunk 1]
```diff
@@ -2523,6 +2523,9 @@ struct EXPORT_VTABLE_ATTRIBUTE Db : NonCopyableClass
         // Skip the check for SQLite file validity before opening. When using the CloudSqlite mode, the local file is not in SQLite normal format.
         bool m_skipFileCheck = false;
 
+        // Provide uri to shared schema channel use by default during schema upgrade or update.
+        Utf8String m_sharedSchemaChannelUri;
```
Comment (kabentley) [inline iModelCore/BeSQLite/PublicAPI/BeSQLite/BeSQLite.h:2527]: what is this uri used for?

Comment (khanaffan) [reply iModelCore/BeSQLite/PublicAPI/BeSQLite/BeSQLite.h:2527]: It's path for cloud sqlite file and parameters like key require to access it.

Comment (kabentley) [reply iModelCore/BeSQLite/PublicAPI/BeSQLite/BeSQLite.h:2527]: that won't work. We need to open this database from JavaScript and pass it in where it's needed.

Comment (khanaffan) [reply iModelCore/BeSQLite/PublicAPI/BeSQLite/BeSQLite.h:2527]: I need to ATTACH the SQLite db to the primary connection. This is the fastest way to sync schemas.

File: iModelCore/BeSQLite/BeSQLite.cpp [hunk 2]
```diff
@@ -5427,11 +5429,14 @@ static void initLanguageSupportOnDb(sqlite3* db)
 +---------------+---------------+---------------+---------------+---------------+------*/
 static int besqlite_db_init(sqlite3* db, char** pzErrMsg, struct sqlite3_api_routines const* pApi)
     {
+
     // and the "InVirtualSet" SQL function. It requires at least two arguments: the address of the VirtualSet and the value(s) to test
-    const auto rc = sqlite3_create_function_v2(db, "InVirtualSet", -1, SQLITE_UTF8, nullptr, &isInVirtualSet, nullptr, nullptr, nullptr);
```
Comment (rschili) [inline iModelCore/BeSQLite/BeSQLite.cpp:5431]: Looks unrelated to the story, but why remove const?
Also, I don't see why we only assert, to me it seems like we should check rc and fail if the function failed to register

File: iModelJsNodeAddon/IModelJsNative.cpp [hunk 3]
```diff
@@ -525,11 +525,37 @@ struct NativeECDb : BeObjectWrap<NativeECDb>, SQLiteOps {
...
         }
     }
+
     Napi::Value SharedChannelEnabled(NapiInfoCR info) {
         RequireDbIsOpen(info);
         const auto isEnabled = !m_ecdb.Schemas().GetSharedChannel().GetInfo().IsEmpty();
         return Napi::Boolean::New(Env(), isEnabled);
     }
+
+    Napi::Value SharedChannelGetLocalInfo(NapiInfoCR info) {
```
Comment (kabentley) [inline iModelJsNodeAddon/IModelJsNative.cpp:535]: again, can we please not call all of this "channels"? We already use that term for access control to models.

Comment (khanaffan) [reply iModelJsNodeAddon/IModelJsNative.cpp:535]: I am renaming everything removing the term shared/channel and use schema sync.

File: iModelCore/BeSQLite/BeSQLite.cpp (hunk 2 above)
Comment (khanaffan) [reply iModelCore/BeSQLite/BeSQLite.cpp:5431]: It's assumed they never fail unless you have an OOM error and subsequent operations will fail anyway. Though we should handle or return an error if the caller provides one from a [branch testability] (/www.tutorialspoint.com/software_testing_dictionary/branch_testing.htm) point of view even if an error has an extremely low chance of ever happening.

File: iModelJsNodeAddon/IModelJsNative.cpp [hunk 4]
```diff
@@ -394,7 +394,10 @@ struct NativeECDb : BeObjectWrap<NativeECDb>, SQLiteOps {
         Napi::HandleScope scope(val.Env());
         return val.As<Napi::Object>().InstanceOf(Constructor().Value());
     }
-
+    void RequireDbIsOpen(NapiInfoCR info) {
```
Comment (kabentley) [inline iModelJsNodeAddon/IModelJsNative.cpp:397]: I don't understand why this is necessary. It's already implemented in SQLiteOps.

File: iModelJsNodeAddon/api_package/ts/src/NativeLibrary.ts [hunk 5]
```diff
@@ -441,9 +441,22 @@ export declare namespace IModelJsNative {
 
   interface SchemaImportOptions {
     readonly schemaLockHeld?: boolean;
+    readonly schemaSyncDbUri?: string;
     readonly ecSchemaXmlContext?: ECSchemaXmlContext;
   }
 
+  interface LocalDbInfo {
```
Comment (kabentley) [inline iModelJsNodeAddon/api_package/ts/src/NativeLibrary.ts:448]: please call this SchemaLocalDbInfo or something
```suggestion
  interface SchemaLocalDbInfo {
```

File: iModelJsNodeAddon/api_package/ts/src/NativeLibrary.ts [hunk 6]
```diff
@@ -441,9 +441,22 @@ export declare namespace IModelJsNative {
...
     readonly ecSchemaXmlContext?: ECSchemaXmlContext;
   }
 
+  interface LocalDbInfo {
+    readonly id: string;
+    readonly dataVer: string;
+    readonly lastModUtc: string;
+  }
+
+  interface SyncDbInfo extends LocalDbInfo {
```
Comment (kabentley) [inline iModelJsNodeAddon/api_package/ts/src/NativeLibrary.ts:454]: SchemaSyncDbInfo

File: iModelCore/ECDb/LiveSync.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
+# No lock schema synchronization across multiple briefcases
+
+No lock does not mean lock free. It just mean that briefcases can make changes to schema without requiring them to push there changes.
```
Comment (kabentley) [inline iModelCore/ECDb/LiveSync.md:3]: it means that there's no need to hold the *schema* lock.

"their", not "there"

File: iModelJsNodeAddon/IModelJsNative.cpp (hunk 4 above)
Comment (khanaffan) [reply iModelJsNodeAddon/IModelJsNative.cpp:397]: We have unsafe calls like the following. An ECDb/iModel that is closed will crash if the following method is called.
```cpp
SchemaManager const& ECDb::Schemas() const { return m_pimpl->Schemas(); }
```

File: iModelCore/ECDb/LiveSync.md (hunk 7 above)
Comment (khanaffan) [reply iModelCore/ECDb/LiveSync.md]: I rewrote it.

File: iModelCore/iModelPlatform/iModelConsole/Command.cpp [hunk 8]
```diff
@@ -269,6 +269,118 @@ void CloseCommand::_Run(Session& session, Utf8StringCR argsUnparsed) const
...
+
+    if (args.empty())
+        {
+        IModelConsole::WriteErrorLine("Usage: %s", GetUsage().c_str());
+        return;
+        }
+
+    if (!session.IsFileLoaded() || session.GetFile().GetECDbHandle()->IsReadonly())
+        {
+        IModelConsole::WriteErrorLine("There is should a file already loaded.");
```
[the remark this answers is not in the document]
Comment (khanaffan) [reply iModelCore/iModelPlatform/iModelConsole/Command.cpp]: fixed

**Review by rschili — APPROVED** [review]

**Review by ArturasDeltuvas — APPROVED** [review]

**Review by kabentley — APPROVED** [review]

File: iModelJsNodeAddon/IModelJsNative.cpp (hunk 4 above)
Comment (kabentley) [reply iModelJsNodeAddon/IModelJsNative.cpp:397]: I understand why you need to call it, but it is already implemented in SQLiteOps. You can just delete this implementation,

Comment (khanaffan) [reply iModelJsNodeAddon/IModelJsNative.cpp]: Fixed.

**Pull request merged**